File: Signal.CPP/WindowFilters.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "MtxExpr.hpp"
#include "WindowFilters.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"
TWindowFiltersForm *WindowFiltersForm;
//---------------------------------------------------------------------------
__fastcall TWindowFiltersForm::TWindowFiltersForm(TComponent* Owner)
	: TForm(Owner)
{
	RichEdit1->Lines->Clear();
	RichEdit1->Lines->Add("Change the length of the filter to see how the width the transition bandwidth changes. "
		"The attenuation in dB applies only to Kaiser in Chebyshev windows. Cutoff frequency "
		"defines the start of the attenuation.");
	FillSeries(0.2,40,128);
}
//---------------------------------------------------------------------------
void __fastcall TWindowFiltersForm::SpectrumEditButtonClick(TObject *Sender)
{
	SpectrumAnalyzerDialog->Execute();
}
//---------------------------------------------------------------------------
void __fastcall TWindowFiltersForm::ChartEditButtonClick(TObject *Sender)
{
	ChartEditor->Execute();
}
//---------------------------------------------------------------------------
void __fastcall TWindowFiltersForm::SpectrumAnalyzerParameterUpdate(TObject *Sender)
{
	SpectrumAnalyzer->Update();
}
//---------------------------------------------------------------------------
void __fastcall TWindowFiltersForm::FillSeries(TSample Freq, TSample Att,int Length)
{
	Mtxexpr::sVector h;
	SpectrumAnalyzer->Window = wtRectangular;
	SpectrumAnalyzer->Logarithmic = true;
	SpectrumAnalyzer->ZeroPadding = 16;
	SpectrumAnalyzer->LogSpan = ls200;
	if (Length % 2 != 0) Length++;
	h->Length = Length;
	for (int i = 0;i<SpectrumChart->SeriesList->Count;i++)
		SpectrumChart->Series[0]->Free();

	for (int i=0;i<11;i++)
	{
		SpectrumChart->AddSeries(new TFastLineSeries(SpectrumChart));
		(*SpectrumChart->Spectrums)[0]->Series = SpectrumChart->Series[i];
		SpectrumChart->Series[i]->Title = SignalWindowToString(TSignalWindowType(i));

		switch ((TSignalWindowType)i)
		{
			case wtBlackman: FirImpulse(h,OPENARRAY(double,(Freq,Freq+0.05)),0.3,ftLowPass, (TSignalWindowType)i,2);break;
			case wtKaiser: FirImpulse(h,OPENARRAY(double,(Freq,Freq+0.05)),KaiserBetaFir(Math387::Power(10,-Att/20)),ftLowPass, (TSignalWindowType)i,2);break;
			default: FirImpulse(h,OPENARRAY(double,(Freq,Freq+0.05)),Att,ftLowPass, (TSignalWindowType)i,2);
		}
		SpectrumAnalyzer->Process(h); //no denominator and place result in self
	}
	SpectrumChart->Legend->Visible = True;


}

void __fastcall TWindowFiltersForm::FreqEditChange(TObject *Sender)
{
	FillSeries(FreqEdit->Position,TransEdit->Position,FilterLengthEdit->IntPosition);
}
//---------------------------------------------------------------------------

File: Signal.Delphi/WindowFilters.pas
unit WindowFilters;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms,
  Dialogs, ComCtrls, ToolWin, ExtCtrls, TeeProcs, TeEngine, Chart,
  SignalToolsTee, TeeEdit, SignalToolsDialogs, SpectrumAna, MtxBaseComp,
  SignalTools, SignalAnalysis, Series, StdCtrls, TeeTools, FileSignal,
  MtxDialogs, MtxComCtrls, Math387, MtxExpr, TeeGDIPlus;

type
  TWindowFiltersForm = class(TForm)
    SpectrumChart: TSpectrumChart;
    Panel1: TPanel;
    ToolBar1: TToolBar;
    SpectrumEditButton: TToolButton;
    ChartEditButton: TToolButton;
    SpectrumAnalyzerDialog: TSpectrumAnalyzerDialog;
    ChartEditor: TChartEditor;
    RichEdit1: TRichEdit;
    ChartTool1: TAxisScaleTool;
    SpectrumAnalyzer: TSpectrumAnalyzer;
    FreqEdit: TMtxFloatEdit;
    Label1: TLabel;
    Label2: TLabel;
    TransEdit: TMtxFloatEdit;
    Label3: TLabel;                                     
    FilterLengthEdit: TMtxFloatEdit;
    TeeGDIPlus1: TTeeGDIPlus;
    procedure SpectrumEditButtonClick(Sender: TObject);
    procedure ChartEditButtonClick(Sender: TObject);
    procedure SpectrumAnalyzerParameterUpdate(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure FreqEditChange(Sender: TObject);
  private
    procedure FillSeries(Freq,Att: TSample; Length: integer);
    { Private declarations }
  public
    { Public declarations }
  end;

var
  WindowFiltersForm: TWindowFiltersForm;

implementation

{$IFDEF CLR}
{$R *.nfm}
{$ELSE}
{$R *.dfm}
{$ENDIF}

uses SignalUtils, MtxVec;

procedure TWindowFiltersForm.SpectrumEditButtonClick(Sender: TObject);
begin
     SpectrumAnalyzerDialog.Execute;
end;

procedure TWindowFiltersForm.ChartEditButtonClick(Sender: TObject);
begin
    ChartEditor.Execute;
end;

procedure TWindowFiltersForm.SpectrumAnalyzerParameterUpdate(Sender: TObject);
begin
     SpectrumAnalyzer.Update;
end;

procedure TWindowFiltersForm.FormCreate(Sender: TObject);
begin
     With RichEdit1.Lines, RichEdit1 do
     begin
        Clear;
        Add('Change the length of the filter to see how the width the transition bandwidth changes. ' +
            'The attenuation in dB applies only to Kaiser in Chebyshev windows. Cutoff frequency '+
            'defines the start of the attenuation.');
        Paragraph.FirstIndent := 0;
        Paragraph.Numbering := nsNone;
     end;
     FillSeries(0.2,40,128);
end;

procedure TWindowFiltersForm.FillSeries(Freq,Att:TSample; Length: integer);
var i: integer;
    H: Vector;
begin
     SpectrumAnalyzer.Window := wtRectangular;
     SpectrumAnalyzer.Logarithmic := True;
     SpectrumAnalyzer.ZeroPadding := 16;
     SpectrumAnalyzer.LogSpan := ls200;
     if Odd(Length) then Inc(Length);
     h.IsDouble := DefaultIsDouble;
     h.Length := Length;
     for i := 0 to SpectrumChart.SeriesList.Count-1 do SpectrumChart.Series[0].Free;
     for i := 0 to 10 do
     begin
         SpectrumChart.AddSeries(TFastLineSeries.Create(SpectrumChart));
         SpectrumChart.Spectrums[0].Series := SpectrumChart.Series[i];
         SpectrumChart.Series[i].Title := SignalWindowToString(TSignalWindowType(i));
         case TSignalWindowType(i) of
         wtBlackman: FirImpulse(h,[Freq,Freq+0.05],0.3,ftLowPass, TSignalWindowType(i),h.Length/2);
         wtKaiser: FirImpulse(h,[Freq,Freq+0.05],KaiserBetaFir(Power(10,-Att/20)),ftLowPass, TSignalWindowType(i),h.Length/2);
         else FirImpulse(h,[Freq,Freq+0.05],Att,ftLowPass, TSignalWindowType(i),h.Length/2);
         end;
         SpectrumAnalyzer.Process(h); //no denominator and place result in self
     end;
     SpectrumChart.Legend.Visible := True;
end;

procedure TWindowFiltersForm.FreqEditChange(Sender: TObject);
begin
     FillSeries(FreqEdit.Position,TransEdit.Position,FilterLengthEdit.IntPosition);
end;

initialization
RegisterClass(TWindowFiltersForm);

end.
